src/pymedphys/dicom/_level3/dicom_collection.py

```diff
@@ -0,0 +1,111 @@
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License and the additional terms for more
+# details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# ADDITIONAL TERMS are also included as allowed by Section 7 of the GNU
+# Affero General Public License. These additional terms are Sections 1, 5,
+# 6, 7, 8, and 9 from the Apache License, Version 2.0 (the "Apache-2.0")
+# where all references to the definition "License" are instead defined to
+# mean the AGPL-3.0+.
+
+# You should have received a copy of the Apache-2.0 along with this
+# program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+
+
+from copy import deepcopy
+
+import pydicom
+
+from .._level1.create import dicom_dataset_from_dict
+from .._level2.anonymise import anonymise_dicom_dataset
+from .._level2.dose import (
+    extract_dicom_patient_xyz, convert_xyz_to_dicom_coords)
+
+from ...libutils import get_imports
+
+IMPORTS = get_imports(globals())
+
+# pylint: disable=W0201
+
+
+class DicomBase:
+    def __init__(self, dataset):
+        self.dataset = dataset
+
+    @property
+    def dataset(self) -> pydicom.Dataset:
+        return deepcopy(self._dataset)
+
+    @dataset.setter
+    def dataset(self, dataset: pydicom.Dataset):
+        self._dataset = deepcopy(dataset)
+        self.after_assigning_dataset()
+
+    def after_assigning_dataset(self):
+        pass
+
+    @classmethod
+    def from_file(cls, filepath):
+        dataset = pydicom.read_file(filepath, force=True)
+
+        return cls(dataset)
+
+    @classmethod
+    def from_dict(cls, dictionary):
+        dataset = dicom_dataset_from_dict(dictionary)
+
+        return cls(dataset)
+
+    def __repr__(self):
+        return self._dataset.__repr__()
+
+    def save_as(self, filepath):
+        self._dataset.save_as(filepath)
+
+    def anonymise(self):
+        anonymised = anonymise_dicom_dataset(self.dataset)
+
+        return self.__class__(anonymised)
+
+
+class DicomDose(DicomBase):
+    def after_assigning_dataset(self):
+        self.values = self._dataset.pixel_array * self._dataset.DoseGridScaling
+        self.units = self._dataset.DoseUnits
+
+        self.x, self.y, self.z = extract_dicom_patient_xyz(self._dataset)
+        self.coords = convert_xyz_to_dicom_coords((self.x, self.y, self.z))
+        self.mask = None
+
+
+class DicomCT(DicomBase):
+    pass
+
+
+class DicomStructure(DicomBase):
+    pass
+
+
+class DicomPlan(DicomBase):
+    pass
+
+
+class DicomCollection:
+    """A DICOM collection can hold one DICOM CT, one DICOM structure, one
+    DICOM plan and one DICOM dose. These DICOM files must align via UID
+    declaration.
+
+    Not all types are required to create a DicomCollection.
+    """
```

Comment:
More thinking required on how best to implement this object...

Reply:
@Centrus007 I guess what I am thinking of here is the case where we have multiple DicomStudy objects which all have pointers to the same DicomSeries and DicomStructure. I don't want the internal dataset object to be duplicated across every DicomStudy. Instead each dataset within the DicomStudy should just be a view.

Each DicomStudy would be a hierarchical object with the DicomSeries (CT) at its base, DicomStructure on top of that, then DicomPlan, then DicomDose.

I envisage a common use case would be `DicomCollection.from_directory(directory_path)`. If there were multiple plan and dose files for just one CT series I would like a DicomStudy for every Plan/Dose combination. That's a lot of DicomStudy instances. I want each DicomSeries within that DicomStudy to only be a pointer. I don't want to use a whole new chunk of RAM for duplicating the DicomSeries data.

It also would be handy to be able to update some feature of the DicomCollection and have every linked DicomObject with the same UID all update in sync. We get this "update in sync" if internally they are actually all pointing to the same address space of RAM.